# Decision trees

- Decision trees are powerful classification and regression models
- Decision trees are trained with labeled data
- App-recommendation, Health care


Decision tree: A machine learning model  based on yes-or-no questions and represented by a binary tree.  The tree has a root node, decision node, left node and branches.

root node: the topmost node of the tree.

Decision node: Each yes-or-no question in our model

Leaf node: A node that has no branch emaning from it.

Depth: the number of levels in the decision tree.


1. Figure out which of the data is the most useful to decide which app to recommend.
1. This feature splits the data into two smaller datasets.
1. Repeat processes 1 and 2 for each of the two smaller datasets.

In other words, what we do is decide which of the two features (platform or age) is more successful at determining which app the users will download and pick this one as our root of the decision tree.

The first step to build our model is to figure out the most useful feature: in other words, the most useful question to ask. First, let’s simplify our data a little bit. Let’s call everyone under 20 years old “Young” and everyone 20 or older “Adult”

In [15]:
import pandas as pd

# Dataset original
data = {
    'Platform': ['iPhone', 'iPhone', 'Android', 'iPhone', 'Android', 'Android'],
    'Age': [15, 25, 32, 35, 12, 14],
    'App': ['Atom Count', 'Check Mate Mate', 'Beehive Finder', 'Check Mate Mate', 'Atom Count', 'Atom Count']
}

df = pd.DataFrame(data)

# Agregar columna 'Age Group'
df['Age_Group'] = df['Age'].apply(lambda age: 'Young' if age < 20 else 'Adult')
print(df)


  Platform  Age              App Age_Group
0   iPhone   15       Atom Count     Young
1   iPhone   25  Check Mate Mate     Adult
2  Android   32   Beehive Finder     Adult
3   iPhone   35  Check Mate Mate     Adult
4  Android   12       Atom Count     Young
5  Android   14       Atom Count     Young


a. Evaluete first question (by platform):

                                            Split by platform
                                           /                 \
                                          iPhone             Android
                                          /                   \
                                Atom Chess Chess             Atom Atom Bee


b. Evaluete first question (by platform):

                                               Split by age
                                           /                 \
                                          Young             Adult
                                          /                   \
                                Atom Atom Atom             Chess Chess Bee


We need to validate which is better questions to split the data.


                            

### Accurracy:  

Accuracy is the fraction of correctly classified data points over the total number of data points.

**Classifier 1**: What platform do you use?


- If the answer is “iPhone,” then we notice that of the iPhone users, the majority downloaded Check Mate Mate. Therefore, we recommend Check Mate Mate to all the iPhone users. We are correct two times out of three
- If the answer is “Android,” then we notice that of the Android users, the majority downloaded Atom Count, so that is the one we recommend to all the Android users. We are correct two times out of three.


**Classifier 2**: What is your age?

- If the answer is “young,” then we notice that all the young people downloaded Atom Count, so that is the recommendation we make. We are correct three times out of three.
- If the answer is “adult,” then we notice that of the adults, the majority downloaded Check Mate Mate, so we recommend that one. We are correct two times out of three.

Notice that classifier 1 is correct four times out of six, and classifier 2 is correct five times out of six. Therefore, for this dataset, classifier 2 is better. 

classifier 1: 4 / 6 = 0.666

classifier 2: 5 / 6 = 0.833


### Gini impurity index: How diverse is my dataset?

Measure of diversity of the dataset.

- if all elements are similar so the gini index is low
- if the elements are diferent so the gini index is is large

For clarity, consider the following two sets of 10 colored balls (where any two balls of the same color are indistinguishable):

Set 1: eight red balls, two blue balls
Set 2: four red balls, three blue balls, two yellow balls, one green ball
Set 1 looks more pure than set 2, because set 1 contains mostly red balls and a couple of blue ones, whereas set 2 has many different colors. Next, we devise a measure of impurity that assigns a low value to set 1 and a high value to set 2. This measure of impurity relies on probability. Consider the following question:

- Set 1: eight red balls, two blue balls
- Set 2: four red balls, three blue balls, two yellow balls, one green ball

Set 1 looks more pure than set 2, because set 1 contains mostly red balls and a couple of blue ones, whereas set 2 has many different colors

If we pick two random elements of the set, what is the probability that they have a different color? The two elements don’t need to be distinct; we are allowed to pick the same element twice.

Note: this is Mutually exclusive events (cannot occur simultaneously,are independent events)

In [16]:
p_once_blue = 8 / 10
p_once_red = 2 / 10

p_twice_blue = p_once_blue ** 2
p_twice_red = p_once_red ** 2


p_diff_color = 1 - p_twice_blue - p_twice_red

# gini impurity index In a set with m elements and n classes,
#  with ai elements belonging to the i-th class, the Gini impurity index is
# Gini = 1 – p1⌃2 – p2⌃2 – … – pn⌃2,


In [17]:
# Gini index to decide which two way to split the data (age or platform)

# Classifier 1: 
# 
# Left leaf: {A, C, C}
# Right leaf: {A, A, B}


# Classifier 2: 
# 
# Left leaf: {A, A, A}
# Right leaf: {B, C, C}

# Checking Left branch of the classifier 1 and classifier 2
# Gini = 1 – p1⌃2 – p2⌃2 – … – pn⌃2,
left_gini_classifier1 = 1 - (2 / 3) ** 2 - (1 / 3) ** 2 
right_gini_classifier1 = 1 - (2 / 3) ** 2 - (1 / 3) ** 2 
avg_gini_classifier1 = 1 / 2 * (left_gini_classifier1 + right_gini_classifier1)

left_gini_classifier2 = 1 - (3 / 3) ** 2
right_gini_classifier2 = 1 - (1 / 3) ** 2 - (2 / 3) ** 2
avg_gini_classifier2 = 1 / 2 * (left_gini_classifier2 + right_gini_classifier2)

print(f"Gini Classifier1 (Platform): {avg_gini_classifier1}")
print(f"Gini Classifier2 (Age): {avg_gini_classifier2}")
print("We conclude that the second split is better, because it has a lower average Gini index.")

Gini Classifier1 (Platform): 0.4444444444444445
Gini Classifier2 (Age): 0.2222222222222222
We conclude that the second split is better, because it has a lower average Gini index.



### Entropy: Another measure of diversity with strong applications in information theory

Consider the same two sets of colored balls as in the previous section, but think of the colors as an ordered set.

- Set 1: {red, red, red, red, red, red, red, red, blue, blue} (eight red balls, two blue balls)
- Set 2: {red, red, red, red, blue, blue, blue, yellow, yellow, green} (four red balls, three blue balls, two yellow balls, one green ball)

### Calculate the probability of get the initial order

**Set1:**

P(r,r,r,r,r,r,r,r,b,b) = 8/10 x 8/10 x 8/10 x 8/10 x 8/10 x 8/10 x 8/10 x 8/10 x 2/10 x 2/10

P(r,r,r,r,r,r,r,r,b,b) = (8/10)⌃8 x (2/10)⌃2

P(r,r,r,r,r,r,r,r,b,b) = 0.006708864

**Set2:**

P(r,r,r,r,b,b,b,y,y,g) =  4/10 x 4/10 x 4/10 x 4/10 x 3/10 * 3/10 x 3/10 x 2/10 x 2/10 x 1 / 10

P(r,r,r,r,b,b,b,y,y,g) =  (4/10)⌃4 x (3/10)⌃4 x  (2/10)⌃4 x (1/10)⌃1  

P(r,r,r,r,r,r,r,r,b,b) = 0.0000027648

The Set1 is more likehook to get the initial order

Why we need Entropy

- Because the probabibilities returns very small decimal and these are cost to process  by the computers


Note: How help log to reduce the complexity : For instance, 0.000000000000001 is equal to 10–15

log(ab) = log(a) + log(b)

log(ac) = c log(a)

Entropy = -1 / n Log base 2 [p1⌃x . p2⌃y . ....]

Entropy = -x /n Log base 2 p1 - y/n log base 2 p2

**Set 1**:

Entropy = - 1 / n Log base 2 [p1⌃x . p2⌃y . ....]

Entropy = -1 / 10 log base 2 [(8/10)⌃8 x (2/10)⌃2]

Entropy = -8 / 10 log base 2 (8/10) - 2 / 10 log base 2 2/10 

Entropy = 0.722

**Set 2**:

Entropy = - 1 / n Log base 2 [p1⌃x . p2⌃y . ....]

Entropy = -1 / 10 log base 2 [(4/10)⌃4 x (3/10)⌃3 x  (2/10)⌃2 x  (1/10)⌃1]

Entropy = -4 / 10 log base 2 (4/10) - 3 / 10 log base 2 3/10 - 2 / 10 log base 2 2/10 - 1 / 10 log base 2 1/10 

Entropy = 1.846

**Notice that the entropy of set 2 is larger than the entropy of set 1, which implies that set 2 is more diverse than set 1. The following is the formal definition of entropy**:

**Entropy = -p1 x log base 2 (p1) - p2 x log base 2 (p2) x .... pn x log base 2 (pn)** 

In [18]:
import math

# Entropy to decide which two way to split the data (age or platform)

# Classifier 1 (Platform): 
# 
# Left leaf: {A, C, C}
# Right leaf: {A, A, B}

c1_p1_left = 2/3
c1_p2_left = 1/3
c1_p1_right = 2/3
c1_p2_right = 1/3

entropy_c1_left = -c1_p1_left *  math.log2(c1_p1_left) - c1_p2_left *  math.log2(c1_p2_left)
entropy_c1_right = -c1_p1_right *  math.log2(c1_p1_right) - c1_p2_right *  math.log2(c1_p2_right)
entropy_c1 = 1 / 2 * (entropy_c1_left + entropy_c1_right)

# Classifier 2 (Age): 
# 
# Left leaf: {A, A, A}
# Right leaf: {B, C, C}


c2_p1_left = 3/3
c2_p1_right = 2/3
c2_p2_right = 1/3

entropy_c2_left = -c2_p1_left *  math.log2(c2_p1_left)
entropy_c2_right = -c2_p1_right *  math.log2(c2_p1_right) - c2_p2_right *  math.log2(c2_p2_right)
entropy_c2 = 1 / 2 * (entropy_c2_left + entropy_c2_right)

print(f"Entropy C1 {entropy_c1}")
print(f"Entropy C2 {entropy_c2}")
print("C2 has less diverse the data so this is better question for us tree node")

Entropy C1 0.9182958340544896
Entropy C2 0.4591479170272448
C2 has less diverse the data so this is better question for us tree node


### Weighted average

El promedio ponderado (o media ponderada) es una forma de calcular un promedio en la que no todos los valores tienen la misma importancia o peso. En lugar de simplemente sumar los valores y dividir entre la cantidad (como en el promedio simple), en el promedio ponderado cada valor se multiplica por un peso que refleja su relevancia, y luego se divide entre la suma total de los pesos.

Imagine that you have a dataset with eight data points (which when training the decision tree, we also refer to as samples), and you split it into two datasets of sizes six and two. As you may imagine, the larger dataset should count for more in the calculations of Gini impurity index or entropy. Therefore, instead of considering the average, we consider the weighted average, where at each leaf, we assign the proportion of points corresponding to that leaf. Thus, in this case, we would weigh the first Gini impurity index (or entropy) by 6/8, and the second one by 2/8.

weight_avg_gini = 0.4444 * 6 / 8  + 0 * 2 / 8 = 0.3333
weight_avg_entropy = 0.918 * 6 /8 + 0 * 2 / 8 = 0.689

### Second step to build the model: Iterating

1. Find the Best Split. Use metrics (like Gini, entropy, or accuracy) to choose the best feature to split the dataset.
1. Split the Dataset. Divide the dataset into subsets based on that feature.
1. Check for Purity
    If all labels in a subset are the same → make it a leaf node with a final prediction.
    If not → go to step 4.
1. Repeat the Process. For each non-pure subset:


### Last step (when stop and hyperparameters)

Since infinite cycles can occur while the tree is being split, we can control it:

1. Don’t split a node if the change in accuracy, Gini index, or entropy is below some threshold.
1. Don’t split a node if it has less than a certain number of samples.
1. Split a node only if both of the resulting leaves contain at least a certain number of samples.
1. Stop building the tree after you reach a certain depth.

#### Hyperparametres

1. The minimum amount of change in accuracy (or Gini index, or entropy)
1. The minimum number of samples that a node must have to split it
1. The minimum number of samples allowed in a leaf node
1. The maximum depth of the tree


### Grid search

The way we pick these hyperparameters is either by experience or by running an exhaustive search where we look for different combinations of hyperparameters and choose the one that performs best in our validation set. This process is called grid search

### Splitting the data using non-binary categorical features, such as dog/cat/bird

 For example, if the input is an animal that could be a dog, a cat, or a bird, then we ask the following questions:

- Is the animal a dog?
- Is the animal a cat?
- Is the animal a bird?

. This process of turning a nonbinary categorical feature into several binary features is called **one-hot encoding** 

Example: 
    dog:   Cat:     bird
      1     0       0 
      0     1       0
      0     0       1

### Splitting the data using continuous features, such as age

Evaluate each posible value as yes/no question

| Platform | Age | App             |
|----------|-----|------------------|
| iPhone   | 15  | Atom Count       |
| iPhone   | 25  | Check Mate Mate  |
| Android  | 32  | Beehive Finder   |
| iPhone   | 35  | Check Mate Mate  |
| Android  | 12  | Atom Count       |
| Android  | 14  | Atom Count       |

Is the user younger than 12?
Is the user younger than 14?
Is the user younger than 15?
Is the user younger than 15?
Is the user younger than 25?
Is the user younger than 32?
Is the user younger than 35?

| Question                          | First set (yes)            | Second set (no)       | Labels (yes), Labels (no)         | Weighted accuracy | Weighted Gini impurity index | Weighted entropy |
|----------------------------------|-----------------------------|------------------------|------------------------------------|-------------------|-------------------------------|------------------|
| Is the user younger than 7?      | empty                      | 12, 14, 15, 25, 32, 35 | {}, {A,A,A,C,B,C}                  | 3/6               | 0.611                         | 1.45             |
| Is the user younger than 13?     | 12                         | 14, 15, 25, 32, 35     | {A}, {A,A,C,B,C}                   | 3/6               | 0.533                         | 1.26             |
| Is the user younger than 20?     | 12, 14, 15                 | 25, 32, 35             | {A,A,A}, {C,B,C}                   | 5/6               | 0.222                         | 0.45             |
| Is the user younger than 28.5?   | 12, 14, 15, 25             | 32, 35                 | {A,A,A,C}, {B,C}                   | 4/6               | 0.416                         | 0.87             |
| Is the user younger than 33.5?   | 12, 14, 15, 25, 32         | 35                     | {A,A,A,C,B}, {C}                   | 4/6               | 0.467                         | 1.14             |
| Is the user younger than 100?    | 12, 14, 15, 25, 32, 35     | empty                  | {A,A,A,C,B,C}, {}                  | 3/6               | 0.611                         | 1.45             |
